push_test/pushclient.py

```diff
@@ -0,0 +1,335 @@
+                await self.process(json.loads(message))
+        except websockets.exceptions.ConnectionClosed:
+            log_msg(out=self.output, status="Websocket Connection closed")
+
+    # Commands:::
+
+    async def _send(self, no_recv=False, **msg):
+        """Send a message out the websocket connection
+
+        :param no_recv: Flag to indicate if response is expected
+        :param msg: message content
+        :return:
+
+        """
+        log_msg(out=self.output, flow="output", msg=msg)
+        await self.connection.send(json.dumps(msg))
+        if no_recv:
+            return
+        message = await self.connection.recv()
+        await self.process(json.loads(message))
+
+    async def cmd_connect(self, server=None, **kwargs):
+        """Connect to a remote websocket server
+
+        :param server: Websocket url
+        :param kwargs: ignored
+        :return:
+
+        """
+        srv = self.config.server or server
+        log_msg(out=self.output, status="Connecting to {}".format(srv))
+        self.connection = await websockets.connect(srv)
+        self.recv.append(asyncio.ensure_future(self.receiver()))
+
+    async def cmd_close(self, **kwargs):
+        """Close the websocket connection (if needed)
+
+        :param kwargs: ignored
+        :return:
+
+        """
+        log_msg(out=self.output, status="Closing socket connection")
+        if self.connection and self.connection.state == 1:
+            try:
+                await self.connection.close()
+                for recv in self.recv:
+                    recv.cancel()
+                self.recv = []
+            except websockets.exceptions.ConnectionClosed:
+                pass
+
+    async def cmd_sleep(self, period=5, **kwargs):
+        log_msg(out=self.output, status="Sleeping...")
+        await asyncio.sleep(period)
+        cmd, args = self._next_task()
+        await getattr(self, cmd)(**args)
+
+    async def cmd_hello(self, uaid=None, **kwargs):
+        """Send a websocket "hello" message
+
+        :param uaid: User Agent ID (if reconnecting)
+
+        """
+        if not self.connection or self.connection.state != 1:
+            await self.cmd_connect()
+        log_msg(out=self.output, status="Sending Hello")
+        args = dict(messageType="hello", use_webpush=1, **kwargs)
+        if uaid:
+            args['uaid'] = uaid
+        elif self.uaid:
+            args['uaid'] = self.uaid
+        await self._send(**args)
+
+    async def cmd_ack(self, channelID=None, version=None,
+                      timeout=60, **kwargs):
+        """Acknowledge a previous mesage
+
+        :param channelID: Channel to acknowledge
+        :param version: Version string for message to acknowledge
+        :param kwargs: Additional optional arguments
+        :param timeout: Time to wait for notifications (used by testing)
+        :return:
+
+        """
+        timeout = timeout * 2
+        while not len(self.notifications):
```

Comment:
If this is a list, then using len isn't needed, it'll evaluate to false when its empty.